### PR TITLE
Fix out-of-bounds read in weight_norm_cuda for empty tensors

### LINKED ISSUE #191848: [CUDA] illegal memory read on weight_norm_fwd_first_dim_kernel
### 🐛 Describe the bug

```python
import torch
torch.cuda.init()
gpu = torch.device('cuda')
v = torch.empty_strided((5, 4, 16, 0), (64, 16, 1, 1), dtype=torch.float32, device=gpu)
g = torch.empty_strided((0, 1, 16, 3), (48, 48, 3, 1), dtype=torch.float32, device=gpu)
torch._weight_norm(v, g, 0)
torch.cuda.synchronize()
```

```
========= COMPUTE-SANITIZER
========= Invalid __global__ read of size 4 bytes
=========     at void at::native::<unnamed>::weight_norm_fwd_first_dim_kernel<float, float>(T1 *, T2 *, const T1 *, const T1 *, int)+0x7e0
=========     by thread (0,0,0) in block (0,0,0)
=========     Address 0x0 is out of bounds
=========     and is 140400297771008 bytes before the nearest allocation at 0x7fb17de00000 of size 1 bytes
```

### Versions

/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 57 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  104
On-line CPU(s) list:                     0-103
Vendor ID:                               GenuineIntel
Model name:                              Intel(R) Xeon(R) Gold 5320 CPU @ 2.20GHz
CPU family:                              6
Model:                                   106
Thread(s) per core:                      2
Core(s) per socket:                      26
Socket(s):                               2
Stepping:                                6
CPU(s) scaling MHz:                      39%
CPU max MHz:                             3400.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4400.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid fsrm md_clear pconfig flush_l1d arch_capabilities
Virtualization:                          VT-x
L1d cache:                               2.4 MiB (52 instances)
L1i cache:                               1.6 MiB (52 instances)
L2 cache:                                65 MiB (52 instances)
L3 cache:                                78 MiB (2 instances)
NUMA node(s):                            2
NUMA node0 CPU(s):                       0-25,52-77
NUMA node1 CPU(s):                       26-51,78-103
Vulnerability Gather data sampling:      Mitigation; Microcode
Vulnerability Indirect target selection: Mitigation; Aligned branch/return thunks
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling:    Not affected
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] nvidia-cublas-cu12==12.6.4.1
[pip3] nvidia-cuda-cupti-cu12==12.6.80
[pip3] nvidia-cuda-nvrtc-cu12==12.6.85
[pip3] nvidia-cuda-runtime-cu12==12.6.77
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cufft-cu12==11.3.0.4
[pip3] nvidia-curand-cu12==10.3.7.77
[pip3] nvidia-cusolver-cu12==11.7.1.2
[pip3] nvidia-cusparse-cu12==12.5.4.2
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-nccl-cu12==2.29.3
[pip3] nvidia-nvjitlink-cu12==12.6.85
[pip3] nvidia-nvtx-cu12==12.6.77
[pip3] torch==2.13.0+cu126
[pip3] triton==3.7.1
[conda] Could not collect

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia

### PR DESCRIPTION
**Pull request by ajaykesarwani**
Fixes #191848

**Summary**
`weight_norm_cuda` (forward and backward) was launching CUDA kernels even when the input tensor `v` has 0 elements (e.g., shape `(5, 4, 16, 0)`). Because the outer dimension is non-zero (5 blocks are launched), each block tries to read from `g[row]` — but `g` has 0 elements, causing an illegal out-of-bounds memory read.

**Reproduction**
```python
import torch
gpu = torch.device('cuda')
v = torch.empty_strided((5, 4, 16, 0), (64, 16, 1, 1), dtype=torch.float32, device=gpu)
g = torch.empty_strided((0, 1, 16, 3), (48, 48, 3, 1), dtype=torch.float32, device=gpu)
torch._weight_norm(v, g, 0)
torch.cuda.synchronize()  # crashes here
```

### CODE REVIEW

**Comment by ajaykesarwani** [comment]
cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia - PTAL when you have a chance! Could someone also approve the workflows so the CI can run?

**Comment by ajaykesarwani** [comment]
Hi @eqy @syed-ahmed @Aidyn-A — following up here. It's been a few days, and the workflows are still awaiting approval, so CI hasn't run yet. Could someone approve the run and/or take a look when you have a moment? Also cc @ptrblck @drisspg to help unblock CI. Thanks for your patience!

**Review by eqy — COMMENTED** [review]
Current protocol is to generalize these types of fixes across multiple backends, are other backends also affected by this issue?

**Comment by eqy** [comment]
@claude review this PR, would other backends also be affected by this issue?

**Comment by ajaykesarwani** [comment]
@eqy Thank you for the detailed review! I've pushed a follow-up commit (`585d76d4`) addressing all the points raised:

- **CPU backend fix**: Added the same `v.numel() == 0` early-return guards to `weight_norm_cpu` and `weight_norm_backward_cpu` in `WeightNorm.cpp`. This also fixes the SIGFPE case (`nn.Linear(4, 0)`).
- **Uninitialized memory**: Added `.zero_()` on `norms` and `grad_g` before the early return in both CUDA and CPU, so the output is deterministic and matches across backends.
- **Tests**: Moved `test_weight_norm_empty_tensor` from `TestNN` to `TestNNDeviceType` for device-generic coverage (CPU, CUDA, MPS, XPU). Rewrote it to assert zeroed values, cover the backward pass, test `dim=-1`, and include the legitimately-shaped empty case (`v=(5,0), g=(5,1)`).

I'll file a follow-up issue for the `TORCH_CHECK` shape validation between `g` and `v` in `_weight_norm` to close the broader family of bugs.

**Comment by ajaykesarwani** [comment]
Hi @eqy, following up — I've pushed 585d76d addressing your point on other backends: CPU had the same bug (and a worse SIGFPE-triggering case), which is now fixed and covered by device-generic tests. Would appreciate a re-review when you have a moment.

Also, CI still hasn't run since the workflows need approval — could you or another maintainer approve the run so we can validate against real test execution? Thanks for your patience!
